src/ir/ExpressionAnalyzer.cpp

```diff
@@ -483,27 +280,47 @@ size_t ExpressionAnalyzer::hash(Expression* curr) {
         // call_imports type, etc. The simplest thing is just to hash the
         // type for all of them.
         rehash(digest, curr->type.getID());
-        // Blocks and loops introduce scoping.
-        if (auto* block = curr->dynCast<Block>()) {
-          noteScopeName(block->name);
-        } else if (auto* loop = curr->dynCast<Loop>()) {
-          noteScopeName(loop->name);
-        } else {
-          // For all other nodes, compare their immediate values
-          visitImmediates(curr, *this);
-        }
-        // Hash children
-        Index counter = 0;
-        for (auto* child : ChildIterator(curr)) {
-          stack.push_back(child);
-          counter++;
-        }
-        // Sometimes children are optional, e.g. return, so we must hash
-        // their number as well.
-        rehash(digest, counter);
+        // Hash the contents of the expression.
+        hashExpression(curr);
       }
     }
 
+    void hashExpression(Expression* curr) {
+
+#define DELEGATE_ID curr->_id
+
+// Create cast versions of it for later operations.
+#define DELEGATE_START(id)                                                     \
+  auto* cast = curr->cast<id>();                                               \
+  WASM_UNUSED(cast);
+
+// Handle each type of field, comparing it appropriately.
+#define DELEGATE_GET_FIELD(id, name) cast->name
+
+#define DELEGATE_FIELD_CHILD(id, name) stack.push_back(cast->name);
+
+#define HASH_FIELD(name) rehash(digest, cast->name);
+
+#define DELEGATE_FIELD_INT(id, name) HASH_FIELD(name)
+#define DELEGATE_FIELD_LITERAL(id, name) HASH_FIELD(name)
+#define DELEGATE_FIELD_SIGNATURE(id, name) HASH_FIELD(name)
+
+#define DELEGATE_FIELD_NAME(id, name) visitNonScopeName(cast->name)
+#define DELEGATE_FIELD_TYPE(id, name) visitType(cast->name);
+#define DELEGATE_FIELD_ADDRESS(id, name) visitAddress(cast->name);
+
+#define DELEGATE_FIELD_SCOPE_NAME_DEF(id, name) noteScopeName(cast->name);
```

Comment:
```suggestion
#define DELEGATE_FIELD_SCOPE_NAME_DEF(id, name)                                \
  noteScopeName(cast->name);                                                   \
  visitScopeName(cast->name);
```
I might be mistaken, but shouldn't we also hash scope names for blocks and loops?

Reply:
I think it's actually better not to. The current code will give the same hash for `(block $x)` (a block with an unneeded name) and `(block)` (with no name). We only give a different hash if there is a noticeable difference in break targets.

But this is a subtle point that I didn't fully think about until now, so it's definitely worth a comment, added.

Reply:
By the way we can't hash expressions with dangling names, which means, something like this, right?
```wasm
(block $label0
  (br $label1)  ;; $label1 is an open name, possibly an outer name
)
```

Is this intentional?

Reply:
It looks like you're right, that wasn't supported before (and doesn't change here). I'm slightly surprised that wasn't an issue. But I guess we don't hash many things atm. Worth fixing eventually.

Reply:
Yeah what I was thinking about was
```wasm
(block $a
  (br $a)
)
```
```wasm
(block $b
  (br $a)
)
```

I was wondering if these two would get the same hash, but actually we weren't even supporting the second case because `$b` was a dangling name. If we want to support dangling names eventually (maybe treat them literally in the same way as we do for comparison), we need to hash block/loop labels too. But given that this hasn't been an issue so far maybe all these are not necessary...

Reply:
Good point, yes, if we support dangling names we'd need to look at this carefully too.